physics/SFC_Models/Land/Noahmp/lnd_iau_mod.F90

```diff
@@ -0,0 +1,848 @@
+!***********************************************************************
+!> TODO: replace with appropriate licence for CCPP
+!*   GNU Lesser General Public License
+!*   <http://www.gnu.org/licenses/>.
+!***********************************************************************
+
+!> @brief Land IAU (Incremental Analysis Update) module, 
+!> for the NoahMP soil/snow temperature (can be extended to include soil moisture) 
+
+!! \section land_iau_mod
+!> - reads settings from namelist file (which indicates if IAU increments are available or not)
+!> - reads in DA increments from GSI/JEDI DA at the start of (the DA) cycle 
+!> - maps increments to FV3 grid points belonging to mpi process
+!> - interpolates temporally (with filter-weights if required by configuration)
+!> - updates states with the interpolated increments
+
+!> March, 2024: Tseganeh Z. Gichamo, (EMC) based on the FV3 IAU mod 
+!> by Xi.Chen <xi.chen@noaa.gov> and Philip Pegion, PSL <philip.pegion@noaa.gov>
+!-------------------------------------------------------------------------------
+
+module land_iau_mod
+  
+  use machine,                  only: kind_phys, kind_dyn
+  use physcons,                 only: pi => con_pi
+  use netcdf
+
+  implicit none
+
+  private
+
+  real(kind=kind_phys), allocatable :: wk3_stc(:, :, :, :), wk3_slc(:, :, :, :)

```

Comment:
@tsga @dustinswales @climbfuji I've made modifications to move the module-defined DDTs to GFS_typedefs (in the FV3 host). I haven't tested/pushed the changes yet. The last bit that could potentially cause a problem is with these variables `wk3_stc, wk3_slc, wk3_slmsk`. They are read in during the init phase and used throughout the simulation in the timestep_init phase, so, basically like LUTs, but they are functions of time and x,y,z. I suppose that it's possible that different instances of the CCPP could read in different files and therefore these variables would be different across instances and therefore would need to get moved to the host as well. @tsga What do you think? Do you think that somewhere down the line (in potentially a different host model), that these variables would be different for each instance?

Reply:
@grantfirl yes, different instances are likely to have different wk3_stc and wk3_slc (but not wk3_slmsk).  Can these also be read by the host at init and passed through to noahmpdrv by the host? (The size of the time dim is often 3, so size of wk3_slc ~ 3*nx*ny*4. z=4 is for soil layers).  Alternatively, I can modify lnd_iau_mod.F90 to read the values from file (during timestep_init) then interpolate and store intermediate values in "Land_IAU_state". BTW, is it correct to assume that Land_IAU_state is not affected by the memory issue?

Reply:
See https://github.com/tsga/ccpp-physics/pull/1. Any module-level variables that need to persist when descoped need to be declared by the host unless those variables can effectively be shared across instances. Since the `wk3*` variables just contain the data read in by the host in the sfc_inc files (correct?), I'm not sure that they would need to be declared by the host. I.e., if different instances read in different sfc_inc files, then, yes, but otherwise, I think no.